Add configurable serializers in each renderer

bf4 opened the pull request:
## Controller Examples

```ruby
class TheController < ApplicationController
  serializing json: ->(json, options) do
    return json if json.is_a?(String)
    json = json.as_json(options) if json.respond_to?(:as_json)

    JSON.pretty_generate(json, options)
  end
end
```


```ruby
class UserController < ApplicationController
  serializing json: ->(json, options) { UserSerializer.new(json, options).as_json(options) }
end
```

## Most relevant core contributor comments

- DHH design feedback https://github.com/rails/rails/pull/21496#r53316942 https://github.com/rails/rails/pull/21496#r53180311
- Rafael Franca approves general implementation following discussion including Jon Moss  https://github.com/rails/rails/pull/21496#issuecomment-220753229 
- Test failures:
  - [ ] https://github.com/rails/rails/pull/21496/files#r55178601 (on backburner until PR is in review again)
- Open question from me:
  - [ ] https://github.com/rails/rails/pull/21496#r55175839

### Comment 2017-09-05

Not sure if it can use JBuilder.  Per https://github.com/rails/jbuilder/issues/321

```ruby
class UserController < ApplicationController
  serializing json: ->(json, options) { 
    template = JbuilderTemplate.new(lookup_context)
    result = JbuilderHandler.call(template)
    JSON.pretty_generate(result)
  }
end
```

### Update 2017-08-30

- Merged in Rails master
- Added relevant core contributor feedback before updates
- Note MissingRenderer https://github.com/rails/rails/pull/21496#r49290079

### Update 2016-09-25

Waiting for further feedback from Rails core. Test failures aren't related.
### Update 2016-05-22

there's been a lot of churn in this PR over the time it's been worked on. Below is the original description.  Current contents is best understood by reading the diff and comments on it.

A good summary can be obtained by reading https://github.com/rails/rails/pull/21496#r53120888 and https://github.com/rails/rails/pull/21496#issuecomment-220753229
### Original description 2015-09-04

per https://groups.google.com/forum/#!topic/rubyonrails-core/K8t4-DZ_DkQ/discussion
- [x] Separate serializer definition from renderer
- [x] Allow specifying `serializer_name` via params
- [x] Changelog
- [ ] Guides 
  - [ ] [guides/source/api_app.md](https://github.com/rails/rails/pull/19832/files#diff-70cc8da393475ab7f80a5625887b0dc9R1)
  - [x] https://github.com/rails/rails/blob/master/guides/source/layouts_and_rendering.md
- [x] clean up tests, perhaps into different files and possibly separate into unit and controller tests
- [x] Agree on implementation
- [ ] raise ActionController::MissingRenderer instead of ActionView::MissingTemplate
- [x] consider which new public methods to introduce
- [x] consider effects on the [responders lib](https://github.com/plataformatec/responders)
- [x] more tests?

Since 2011, work in EmberJS and on the JSON API has resulted in some changing naming conventions
- Model:  class that defines the properties and behavior of the data that you present to the user.
- Record: A record is an instance of a model that contains data loaded from a server. Your application can also create new records and save them back to the server.
- Adapter: knows how to talk to your server. knows how to translate requests from the client into requests on your server. object that translates requests from AMS/Rails/Grape etc (such as "find the user with an ID of 123") into a requests to a server. let you completely change how your API is implemented without impacting your application code.
- Serializer: maps keys and values to desired format, see https://github.com/orbitjs/orbit.js/blob/master/lib/orbit-common/jsonapi-serializer.js
- Store: the central repository of records in your application, use the store to retrieve records, as well to create new ones. The store will automatically cache records for you.

bf4 committed:
Add configurable serializers in each renderer

(This commit will be cleaned up. It just makes rebasing easier).

Squashed from

b0bc25d4c01012341515d635775c151fb786e8e0 to

e355bdb83726b817151b5ecaa8b6438986a84ba9

Respond to comments, not amended for ease of reviewing

Separate serializers from renderers

Pass in serializer_name as param

Refactor _render_to_body_with_renderer

Consider what from Responders should be in here

Bring parity to _serializers

Write up CHANGELOG entry

Fix failing tests

Respond to feedback by rafael franca

bf4 commented:
Existing tests should ensure no regression.  If we move forward with this design or other feedback, I'll add tests and documentation for the new features as well.

bf4 commented:
r? @guilleiguaran

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +123,88 @@ def self.remove(key)
+          self._serializers = serializers.freeze
+        end
+        alias use_serializer use_serializers
+      end
```
might not need this, but I haven't looked into this vector, yet

bf4 commented:
cc @joaomdmoura

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -62,12 +78,16 @@ def self._render_with_renderer_method_name(key)
     #
     # Create a csv renderer:
     #
+    #   ActionController::Serializers.add :csv do |obj, options|
```
I thkik this is wrong.

rafaelfranca replied:
Does we have a method `add` in `ActionController::Serializers`?

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +123,88 @@ def self.remove(key)
     module All
       extend ActiveSupport::Concern
       include Renderers
+      include Renderers::Serializers
```
Seems weird to this module be inside `Renderers`

rafaelfranca commented:
API looks weird but implementation is good. Maybe it would be better if we don't have the `Serializers` module and only the methods inside the `Renderers` module.

bf4 commented:
Thanks @rafaelfranca 

>  only the methods inside the Renderers module.

That works for me, and probably makes more sense and is simpler to understand. 

I also apologize that I don't seem to have noted that this is a proposed solution that doesn't break anything (per tests), and that if we go with it, I'll write tests and some docs.  So, I especially thank your for  treatment of a PR like this without tests :)

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +123,88 @@ def self.remove(key)
+
     add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+      json = _serialize_with_serializer_json(json, options)
```
Although I'm not a *huge* fan of this interface, I think it has the following benefits

1. It serves to implicitly document the use of methods like `_serialize_with_serializer_json`, which makes them easier to search for
1. It's a very small change
1. I think it could be a good thing to separate the ides of a 'json renderer' that uses as 'json serializer', yet keep them tightly coupled in usage and code location, even though there is a slightly higher cognitive burden.  Perhaps we should ensure the renderer has a helpful failure message if the serializer isn't defined.

bf4 replied:
Maybe would be nicer to have a sort of serializer middleware, so that we could extend it? e.g. ActiveModel::Serializers transforms it to `as_json` and another layer takes calls `to_json(options)` on that or `JSON.pretty_generate(json, options)`

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +123,88 @@ def self.remove(key)
+    # +ActionController::Serializers.add_serializer json do |json, options|+.
+    #
+    # See https://groups.google.com/forum/#!topic/rubyonrails-core/K8t4-DZ_DkQ/discussion for
+    # more background information.
```
What do you think of this documentation?

```ruby
    # Serializers define a method called within a renderer specific to
    # transforming the object into a mime-compatible type.
    #
    # The separation of serialization from rendering allows
    # composing the Renderer behavior of two methods, e.g.
    # +_render_with_renderer_json+ and +_serialize_with_serializer_json+,
    # rather than requiring one to define a method +_render_with_renderer_json+
    # in a subclass and optionally call super on it.
    #
    # A principal benefit of this approach is that it promotes serialization on an object
    # to a clearly-defined public interface, rather than requiring one to understand that
    # calling, e.g. +render json: object+ calls +_render_to_body_with_renderer(options)+
    # which calls +_render_with_renderer_#{key}+ where key is +json+, which is the method
    # defined by calling +ActionController::Renderers.add :json+.
    #
    # ActiveModel::Serializers, for example, which has relied upon defining
    # +_render_option_json+ and +_render_with_renderer_json+ in the controller, and calling +super+
    # on the serialized object, now only needs to call
    # +ActionController::Serializers.remove_serializer :json+ and define a new serializer with
    # +ActionController::Serializers.add_serializer json do |json, options|+.
    #
    # See https://groups.google.com/forum/#!topic/rubyonrails-core/K8t4-DZ_DkQ/discussion for
    # more background information.
```

bf4 commented:
@rafaelfranca  Updated 

cc 
- @fxn https://github.com/rails/rails/commit/4dfe140ef3fbed34e6ae8af3b32402630a4690a1
- @amatsuda https://speakerdeck.com/a_matsuda/render-it-a-deep-dive-into-actionview-and-template-engines
- @zzak https://github.com/rails/rails/commit/f7ebdb1ac51f26cff76e5642a75717df1b446746

I figure tests might go in (?)
- https://github.com/rails/rails/blob/master/actionpack/test/controller/renderer_test.rb
- https://github.com/rails/rails/blob/master/actionpack/test/controller/render_xml_test.rb

and doc changes would go in

- https://github.com/rails/rails/blob/master/guides/source/layouts_and_rendering.md
- https://github.com/rails/rails/blob/master/actionpack/CHANGELOG.md

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +142,106 @@ def self.remove(key)
+    #     json = CustomSerializer.new(json, options)
+    #     super
+    #   end
+    #
```
It's interesting that you can't really look up a renderer/serializer right now, that the dance Rails is doing is mapping e.g. `:json` to `_render_with_renderer_json(json, options)`, e.g. that to [render to json, it iterates, builds a method name, and sends to it](https://github.com/rails/rails/blob/8c2c1bccccb3ea1f22503a184429d94193a1d9aa/actionpack/lib/action_controller/metal/renderers.rb#L44-L45)

```ruby
    def render_to_body(options)
      _render_to_body_with_renderer(options) || super
    end

    def _render_to_body_with_renderer(options)
      _renderers.each do |name|
        if options.key?(name)
          _process_options(options)
          method_name = Renderers._render_with_renderer_method_name(name)
          return send(method_name, options.delete(name), options)
        end
      end
      nil
    end
```

That there's no `Renderers.for(:json).render(json, options)` which might use an object like `JsonRenderer` or even the proc/block defined as the body of `_render_with_renderer_json`.

And having [ActionController::Rendering.renderer](https://github.com/rails/rails/blob/fecf9d729a81c6c08fef92936e659cdfcf340be5/actionpack/lib/action_controller/metal/rendering.rb#L16-L18) call [ActionController::Renderer.for(controller)](https://github.com/rails/rails/blob/3e36db4406beea32772b1db1e9a16cc1e8aea14c/actionpack/lib/action_controller/renderer.rb#L42-L54) which creates and anonymous subclass of `ActionController::Renderer` it appears is only used for templates?

rafaelfranca replied:
Yeah, this API is something we can improve in a new PR

bf4 commented:
Is there anything I can do to help move this along?

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -11,28 +11,61 @@ def self.remove_renderer(key)
   end
 
+  # RFC: Where is this used?
   class MissingRenderer < LoadError
```
I think it is not being used

rafaelfranca replied:
Actually it is on responders https://github.com/plataformatec/responders/blob/f0e093d01e799d8b67952c013daf8808e9ea02db/lib/action_controller/responder.rb#L209.

bf4 replied:
good spleunking.  Now, the code over there is

```ruby
    # This is the common behavior for formats associated with APIs, such as :xml and :json.
    def api_behavior
      raise MissingRenderer.new(format) unless has_renderer?
      # etc
    end

    # Check whether the necessary Renderer is available
    def has_renderer?
      Renderers::RENDERERS.include?(format)
    end
```

which makes me think either the exception should be defined in responders  or `has_renderer?` should be defined here, as well.

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -11,28 +11,61 @@ def self.remove_renderer(key)
     end
 
     module ClassMethods
+      # RFC: Where is this used?
```
This can be used when the controller don't include `Renderers::All`

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +142,106 @@ def self.remove(key)
+    #
+    # Example usage:
+    #
+    # Prior to the introduction of SERIALIZERS, the ActiveModel::Serializers gem,
```
I'd not talk about the gem here. The code examples are fine.

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +142,106 @@ def self.remove(key)
+    #
+    # The JSON serializer could be extended by overriding it:
+    #
+    #   def _serialize_with_serializer_json(json, options)
```
Why do you would want to do this if you can remove and add a new serializer?

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +142,106 @@ def self.remove(key)
+    #     super
+    #   end
+    #
+    # The JSON serializer could be extended by overriding the renderer:
```
overriding the renderer if you want to extend the serializer?

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -103,17 +142,106 @@ def self.remove(key)
+    #     super
+    #   end
+    #
+    # The JSON serializer could be extended by re-using the existing serializer:
```
I think we don't need this example

rafaelfranca commented:
> figure tests might go in (?)
> 
> https://github.com/rails/rails/blob/master/actionpack/test/controller/renderer_test.rb
> https://github.com/rails/rails/blob/master/actionpack/test/controller/render_xml_test.rb
> https://github.com/rails/rails/blob/master/actionpack/test/controller/mime/respond_to_test.rb#L139
> https://github.com/rails/rails/blob/master/actionpack/test/controller/render_json_test.rb#L90
> it looks like some tests may have been removed in ee77770#diff-28658021f35e96c7499fdf41908629f0L644 ? https://github.com/rails/rails/blob/aee1ba42339df2baf6b44af01f270ae2d37863e1/actionpack/test/controller/mime/respond_with_test.rb#L644-L678

:+1:

joaomdmoura commented:
:+1:

bf4 commented:
cc @tenderlove @sgrif

bf4 commented:
I'm having trouble pushing this right now, but here's the new diff

```diff
diff --git a/actionpack/lib/action_controller/metal/renderers.rb b/actionpack/lib/action_controller/metal/renderers.rb
index 531e6c7..f6ac3f1 100644
--- a/actionpack/lib/action_controller/metal/renderers.rb
+++ b/actionpack/lib/action_controller/metal/renderers.rb
@@ -11,6 +11,7 @@ module ActionController
     Renderers.remove(key)
   end
 
+  # TODO: update text
   # See <tt>Responder#api_behavior</tt>
   #
   #   # This is the common behavior for formats associated with APIs, such as :xml and :json.
@@ -86,7 +87,12 @@ module ActionController
         if options.key?(name)
           _process_options(options)
           method_name = Renderers._render_with_renderer_method_name(name)
-          return send(method_name, options.delete(name), options)
+          value = options.delete(name)
+          serializer_name = options.delete(:serializer_name) || name
+          serializer = Renderers::SERIALIZERS[serializer_name] or
+            fail ActionController::MissingSerializer, "There is no '#{serializer_name}' serializer for '#{name}'"
+          value = serializer.call(value, options)
+          return send(method_name, value, options)
         end
       end
       nil
@@ -153,14 +159,12 @@ module ActionController
       remove_method(method_name) if method_defined?(method_name)
     end
 
-    # A Set containing serializer names that correspond to available serializer procs.
-    # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
-    SERIALIZERS = Set.new
-
-    def self._serialize_with_serializer_method_name(key)
-      "_serialize_with_serializer_#{key}"
-    end
+    # A Hash mapping serializer names to callables.
+    # TODO: update text
+    # Default serializers are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
+    SERIALIZERS = Hash.new
 
+    # TODO: update text
     # Serializers define a method called within a renderer specific to
     # transforming the object into a mime-compatible type.
     # See <tt>Renderers.add</tt>
@@ -200,8 +204,7 @@ module ActionController
     # See https://groups.google.com/forum/#!topic/rubyonrails-core/K8t4-DZ_DkQ/discussion for
     # more background information.
     def self.add_serializer(key, &block)
-      define_method(_serialize_with_serializer_method_name(key), &block)
-      SERIALIZERS << key.to_sym
+      SERIALIZERS[key.to_sym] = block
     end
 
     # This method is the opposite of add_serializer method.
@@ -211,8 +214,6 @@ module ActionController
     #   ActionController.remove_serializer(:csv)
     def self.remove_serializer(key)
       SERIALIZERS.delete(key.to_sym)
-      method_name = _serialize_with_serializer_method_name(key)
-      remove_method(method_name) if method_defined?(method_name)
     end
 
     module All
@@ -230,8 +231,6 @@ module ActionController
     end
 
     add :json do |json, options|
-      json = _serialize_with_serializer_json(json, options)
-
       if options[:callback].present?
         if content_type.nil? || content_type == Mime[:json]
           self.content_type = Mime[:js]
@@ -250,12 +249,10 @@ module ActionController
 
     add :js do |js, options|
       self.content_type ||= Mime[:js]
-      _serialize_with_serializer_js(js, options)
     end
 
     add :xml do |xml, options|
       self.content_type ||= Mime[:xml]
-      _serialize_with_serializer_xml(xml, options)
     end
 
     add_serializer :xml do |xml, options|
diff --git a/actionpack/test/controller/mime/respond_to_test.rb b/actionpack/test/controller/mime/respond_to_test.rb
index c025c7f..615c3e3 100644
--- a/actionpack/test/controller/mime/respond_to_test.rb
+++ b/actionpack/test/controller/mime/respond_to_test.rb
@@ -136,6 +136,7 @@ class RespondToController < ActionController::Base
     end
   end
 
+  # wip
   def json_with_callback
     respond_to do |type|
       type.json { render :json => 'JS', :callback => 'alert' }
diff --git a/actionpack/test/controller/render_json_test.rb b/actionpack/test/controller/render_json_test.rb
index 3773900..8c3c4b7 100644
--- a/actionpack/test/controller/render_json_test.rb
+++ b/actionpack/test/controller/render_json_test.rb
@@ -86,7 +86,7 @@ class RenderJsonTest < ActionController::TestCase
     assert_equal '[]', @response.body
   end
 
-
+  # wip
   def test_render_json
     get :render_json_hello_world
     assert_equal '{"hello":"world"}', @response.body
diff --git a/actionpack/test/controller/render_xml_test.rb b/actionpack/test/controller/render_xml_test.rb
index f0fd7dd..79f5db0 100644
--- a/actionpack/test/controller/render_xml_test.rb
+++ b/actionpack/test/controller/render_xml_test.rb
@@ -3,6 +3,7 @@ require 'controller/fake_models'
 require 'pathname'
 
 class RenderXmlTest < ActionController::TestCase
+  # wip
   class XmlRenderable
     def to_xml(options)
       options[:root] ||= "i-am-xml"
diff --git a/actionpack/test/controller/renderer_test.rb b/actionpack/test/controller/renderer_test.rb
index 16d24fa..c48d61c 100644
--- a/actionpack/test/controller/renderer_test.rb
+++ b/actionpack/test/controller/renderer_test.rb
@@ -1,6 +1,7 @@
 require 'abstract_unit'
 
 class RendererTest < ActiveSupport::TestCase
+  # wip
   test 'action controller base has a renderer' do
     assert ActionController::Base.renderer
   end
diff --git a/actionpack/test/controller/renderers_test.rb b/actionpack/test/controller/renderers_test.rb
new file mode 100644
index 0000000..fe37626
--- /dev/null
+++ b/actionpack/test/controller/renderers_test.rb
@@ -0,0 +1,100 @@
+require 'abstract_unit'
+require 'controller/fake_models'
+require 'active_support/logger'
+
+class RenderersTest < ActionController::TestCase
+  class TestController < ActionController::Base
+    class XmlRenderable
+      def to_xml(options)
+        options[:root] ||= "i-am-xml"
+        "<#{options[:root]}/>"
+      end
+    end
+    class JsonRenderable
+      def as_json(options={})
+        hash = { :a => :b, :c => :d, :e => :f }
+        hash.except!(*options[:except]) if options[:except]
+        hash
+      end
+
+      def to_json(options = {})
+        super :except => [:c, :e]
+      end
+    end
+    class CsvRenderable
+      def to_csv
+        "c,s,v"
+      end
+    end
+
+    def respond_to_mime
+      respond_to do |type|
+        type.json { render json: JsonRenderable.new    }
+        type.js   { render json: 'JS', callback: 'alert' }
+        type.csv  { render csv: CsvRenderable.new    }
+        type.xml  { render xml: XmlRenderable.new     }
+        type.html { render body: "HTML"    }
+        type.rss  { render body: "RSS"     }
+        type.all  { render body: "Nothing" }
+        type.any(:js, :xml) { render body: "Either JS or XML" }
+      end
+    end
+  end
+
+  tests TestController
+
+  def setup
+    # enable a logger so that (e.g.) the benchmarking stuff runs, so we can get
+    # a more accurate simulation of what happens in "real life".
+    super
+    @controller.logger = ActiveSupport::Logger.new(nil)
+  end
+
+  def test_raises_missing_template_no_renderer
+    assert_raise ActionView::MissingTemplate do
+      get :respond_to_mime, format: 'csv'
+    end
+    assert_equal Mime[:csv], @response.content_type
+    assert_equal "", @response.body
+  end
+
+  def test_adding_csv_rendering
+    ActionController::Renderers.add :csv do |value, options|
+      send_data value, type: Mime[:csv]
+    end
+    ActionController::Renderers.add_serializer :csv do |value, options|
+      value.to_csv
+    end
+    @request.accept = "text/csv"
+    get :respond_to_mime, format: 'csv'
+    assert_equal Mime[:csv], @response.content_type
+    assert_equal "c,s,v", @response.body
+  ensure
+    ActionController::Renderers.remove :csv
+  end
+
+  def test_replacing_serializer
+    default_json_serializer = ActionController::Renderers::SERIALIZERS[:json]
+    get :respond_to_mime, format: 'json'
+    assert_equal TestController::JsonRenderable.new.to_json, @response.body
+
+    ActionController::Renderers.remove_serializer :json
+
+    assert_raise ActionController::MissingSerializer do
+      get :respond_to_mime, format: 'json'
+    end
+
+    ActionController::Renderers.add_serializer :json do |json, options|
+      return json if json.is_a?(String)
+
+      json = json.as_json(options) if json.respond_to?(:as_json)
+      json = JSON.pretty_generate(json, options)
+      json
+    end
+    get :respond_to_mime, format: 'json'
+    assert_equal JSON.pretty_generate(TestController::JsonRenderable.new.as_json), @response.body
+  ensure
+    ActionController::Renderers::SERIALIZERS[:json] = default_json_serializer
+  end
+
+end
```

bf4 commented:
Having trouble passing through `serializer_name`

bf4 commented:
Put `serializer_name` (which is up for discussion) in `params` hash with @sgrif

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -23,27 +54,44 @@ module Renderers
+        assert_serializer!(serializer_name)
+        serializer = serializer_for(serializer_name)
+        value = serializer.call(value, options)
+        return send(method_name, value, options)
```
TODO: raise ActionController::MissingRenderer instead of ActionView::MissingTemplate

bf4 commented on actionpack/test/controller/renderers_test.rb:
```diff
@@ -0,0 +1,125 @@
+  end
+
+  def test_raises_missing_template_no_renderer
+    assert_raise ActionView::MissingTemplate do
```
TODO: replace with ActionController::MissingRenderer

bf4 commented on actionpack/test/controller/renderers_test.rb:
```diff
@@ -0,0 +1,125 @@
+    get :respond_to_mime, format: 'json'
+    assert_equal JSON.pretty_generate(JsonRenderable.new.as_json), @response.body
+  ensure
+    ActionController::Renderers::SERIALIZERS[:json] = default_json_serializer
```
I don't want to recommend this usage, but it seems more direct here.

bf4 commented:
cc @spastorino this serializers PR and #22519, k thanks :)

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -65,19 +98,49 @@ def use_renderers(*args)
+      return true if has_serializer?(serializer_name)
+      raise ActionController::MissingSerializer, "There is no '#{serializer_name}' serializer.\n" <<
+            "Known serializers are #{_serializers.keys}"
+    end
```
Funny story.  [ActiveModelSerializers had an `assert_serializer` test helper that was removed, and has a PR to re-add it ready to be merged](https://github.com/rails-api/active_model_serializers/pull/1390).  I'll need to think about the differences in what each gives.  The AMS one is more concerned with the serializer class, not the rendering serializer. i.e.` assert_serializer(PostSerializer)` vs. `assert_serializer!(:json)`

rafaelfranca replied:
This method is being called only in one place, does it make sense to have it?

bf4 replied:
Only that it makes the `_render_to_body_with_renderer` method a little more manageable

```diff
    def _render_to_body_with_renderer(options)
      _renderers.each do |name|
-        if options.key?(name)
-          _process_options(options)
-          method_name = Renderers._render_with_renderer_method_name(name)
-          return send(method_name, options.delete(name), options)
-        end
+        next unless options.key?(name)
+        _process_options(options)
+        method_name = Renderers._render_with_renderer_method_name(name)
+        value = options.delete(name)
+        serializer_name = options.key?(:serializer_name) ? options.delete(:serializer_name) : name
+        assert_serializer!(serializer_name)
+        serializer = serializer_for(serializer_name)
+        value = serializer.call(value, options)
+        return send(method_name, value, options)
      end
      nil
    end
```

bf4 commented:
@rafaelfranca I extracted the controller renderers test into https://github.com/rails/rails/pull/22906 so we can diff any changes to it causes by this PR.

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -16,13 +16,65 @@ def initialize(format)
+  #   # Check whether the necessary Renderer is available
+  #   def has_renderer?
+  #     Renderers::RENDERERS.include?(format)
+  #   end
```
I don't think there's really an easy way to do implement `MissingRenderer` at this time. Given this file defines


```ruby
     def render_to_body(options)
        _render_to_body_with_renderer(options) || super
      end
```

When there's no registered renderer,`_render_to_body_with_renderer(options)` is `nil` and the method calls `super`, which may raise `ActionView::MissingTemplate` in an ActionController::Base, but just be nil in an ActionController::Metal. So, if we consider a renderer an `API` concern, in the language of the [`Responders` gem](https://github.com/plataformatec/responders/blob/f0e093d01e799d8b67952c013daf8808e9ea02db/lib/action_controller/responder.rb#L209), how do we know which `content_type` or `Mime::Type.lookup(content_type)` should be a missing template (html/plain/text) vs. missing renderer (json/xml/js/csv)?

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -16,13 +16,65 @@ def initialize(format)
 
+    # A Set containing renderer names that correspond to available renderer procs.
+    # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
+    RENDERERS = Set.new
```
Was thinking an interesting style would be to define renderers per mime_type or format e.g. it would build something like

Given:

```ruby
content_type = 'application/json'
mime_type = Mime::Type.lookup(content_type)
format = :json
default_renderer = :json
```

then

```ruby
RENDERERS = {
  json: {  # 'format', e.g. `render json: model`
    content_type: Mime[:json], # each format has an associated mime type
    serializers: {
      default: "_render_with_renderer_json(json, options)" # this is basically the default json renderer.  Maybe it should be a callable, but this is the idea
    }
  }
}
```

Then, you could `ActionController::Renderers.add :json, :custom_serializer { |json, options| whatever }` and `render json: model, serializer: :custom_serializer` or maybe even `use_serializer json: :custom_serializer`.

To render JSON API it might make sense to add a `json_api` renderer and serializer e.g.

```ruby
RENDERERS = {
  json: {  } # etc, see above
  json_api: {  # 'format', e.g. `render json_api: model`
    content_type: Mime[:json-api], # each format has an associated mime type
    serializers: {
      default: "ActiveModel::SerializableResource.new(json, options)" # this is basically the ActiveModelSerializers serializer
    }
  }
}
```

But I'm not sure I'd want to change my renders from the `render json: model`, so maybe consider that the `json` format, and it uses the `json` renderer (calls `as_json` and sets `content_type`) and `serializer: :json_api` (representation of 'model' that be as_json, if not a string)

dhh replied:
Here's a simple proposal for cutting down on the specialness of the configuration as much as possible:

```ruby
class ApplicationController < ActionController::Base
  serializing json: ->(object, options) { ActiveModel::SerializableResource.new(model, options) },
               xml: ->(object, options) { XmlSerializer.new(object, options) }
end

class PostsController < ApplicationController
  def show
    @post = Post.find(params[:id])
    
    respond_to do |format|
      format.json { render json: @post }
      format.xml  { render  xml: @post }
      format.csv  { render  csv: CsvSerializer.serialize(@post) }
    end
  end
end
```

Then we don't need any external registry idea. You can simply overwrite the serializing function on a per-controller basis, or in ApplicationController, if you want to set a default. This would then also mean that we constrain serializing overwrites to ApplicationController an up, which would mean this wouldn't fuck with an engine might expect.

bf4 replied:
Follow up comment in https://github.com/rails/rails/pull/21496/files#r54989781, WIP pr being pushed while I think about implementation. (I haven't gotten it 'right', yet.)

bf4 commented on actionpack/test/controller/renderers_test.rb:
```diff
@@ -56,16 +60,20 @@ def setup
+    ActionController.remove_serializer :simon
   end
 
   def test_raises_missing_template_no_renderer
```
PR NOTE: Just some comments about MissingRenderer perhaps for a future PR

I don't think there's really an easy way to do implement `MissingRenderer` at this time. Given
this file defines


```ruby
def render_to_body(options)
  _render_to_body_with_renderer(options) || super
end
```

When there's no registered renderer,`_render_to_body_with_renderer(options)`
is `nil` and the method calls `super`, which may raise
`ActionView::MissingTemplate` in an ActionController::Base, but just be `nil`
in an ActionController::Metal. So, if we consider a renderer an `API`
concern, in the language of the [`Responders`
gem](https://github.com/plataformatec/responders/blob/f0e093d01e799d8b67952c013daf8808e9ea02db/lib/action_controller/responder.rb#L209),
how do we know which `content_type` or `Mime::Type.lookup(content_type)`
should be a missing template (html/plain/text) vs. missing renderer
(json/xml/js/csv)?

Responders implementation for reference:

```ruby
  # This is the common behavior for formats associated with APIs, such as :xml and :json.
  def api_behavior
    raise MissingRenderer.new(format) unless has_renderer?
  end

  # Check whether the necessary Renderer is available
  def has_renderer?
    Renderers::RENDERERS.include?(format)
  end
```

bf4 replied:
ref: https://github.com/rails/rails/pull/23564

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -25,9 +42,16 @@ module Renderers
     # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
     RENDERERS = Set.new
```
Was thinking an interesting style would be to define renderers per mime_type or format e.g. it would be something like

Given:

```ruby
content_type = 'application/json'
mime_type = Mime::Type.lookup(content_type)
format = :json # we can't rely on a format passed to render nor that it corresponds to a mime type.
default_renderer = :json
```

then

```ruby
RENDERERS = {
  json: {  # 'format', e.g. `render json: model`
    content_type: Mime[:json], # each format has an associated mime type
    adapters: { # I called this a serializer in the original comment, but am thinking adapter might be better?
      default: "_render_with_renderer_json(json, options)" # this is basically the default json renderer.  Maybe it should be a callable, but this is the idea
    }
  }
}
```

Then, you could `ActionController::Renderers.add :json, custom_adapter: { |json, options| whatever }` and `render json: model, adapter: :custom_adapter` or maybe even `use_adapter json: :custom_adapter`.

To render JSON API it might make sense to add a `json_api` renderer and adapter e.g.

```ruby
RENDERERS = {
  json: {  } # etc, see above
  json_api: {  # 'format', e.g. `render json_api: model`
    content_type: Mime[:json_api], # each format has an associated mime type
    adapters: {
      default: "ActiveModel::SerializableResource.new(json, options)" # this is basically the ActiveModelSerializers serializer
    }
  }
}
```

But I'm not sure I'd want to change my renders from the `render json: model`, so maybe consider that the `json` format, and it uses the `json` renderer (calls `as_json` and sets `content_type`) and `adapter: :json_api` (representation of 'model' that be as_json, if not a string)

```ruby
RENDERERS = {
  json: {  # 'format', e.g. `render json: model`
    adapters: {
      default: {
        content_type: Mime[:json], # each serializer has an associated mime type
        adapter: "_render_with_renderer_json(json, options)" # this is basically the default json renderer.  Maybe it should be a callable, but this is the idea
      },
      json_api:  {
        content_type: Mime[:json_api], # each serializer has an associated mime type
        adapter: "ActiveModel::SerializableResource.new(json, options)" # this is basically the ActiveModelSerializers serializer
      },
    }
  }
}
```

then we could

```ruby
render json: model
# OR
render json: model, adapter: :default
# OR
render json: model, adapter: :json_api
# OR
use_renderer json: :json_api
def show
  render json: model #=== render json: model, adapter: :json_api
end
## AND
use_renderer :json
def show
  render json: model #=== render json: model, adapter: :default
end
```

rafaelfranca replied:
I think this idea. @dhh WDYT about this API?

dhh replied:
Let me give this some thought. Definitely want to make some amendments here.

bf4 replied:
Thanks. I've decided I don't like the impl with the renderer and serializer
being one-to-one, but haven't settled on anything better, so those ideas
are still just in comments. I feel like I still haven't clarified what's
missing from the example impl. I think part of it is the role responders
played. The more minds the better!
On Wed, Feb 17, 2016 at 9:34 AM David Heinemeier Hansson <
notifications@github.com> wrote:

> In actionpack/lib/action_controller/metal/renderers.rb
> <https://github.com/rails/rails/pull/21496#discussion_r53180311>:
>
> > @@ -25,9 +42,16 @@ module Renderers
> >      # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
> >      RENDERERS = Set.new
>
> Let me give this some thought. Definitely want to make some amendments
> here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/21496/files#r53180311>.
>

bf4 replied:
A small update here, with respect to JSON API and discussion in https://github.com/rails/rails/pull/23712, that we're going to make a distinction between different renderers and rendering different media types.

Longer discourse on how this plays out:

## Different renderers for rendering different media types

e.g. JSON API has a different mime-type from `application/json`, so should have its own renderer, such that: 
- `respond_to do |format| format.jsonapi { render jsonapi: model } end` 
  - i.e. when `request.format == Mime[:jsonapi]`, then it will use the `Renderers::RENDERERS[:jsonapi]` and return `response.content_type = Mime[:jsonapi]`

It would be a *bug* for `render json:` to respond to a request with the JSON API media type and not render a JSON API document nor set the proper response content_type.

```ruby
  MEDIA_TYPE = 'application/vnd.api+json'
  HEADERS = {
    response: { 'CONTENT_TYPE' => MEDIA_TYPE },
    request:  { 'ACCEPT' => MEDIA_TYPE }
  }
 Mime::Type.register MEDIA_TYPE, :jsonapi
 media_type = Mime[:jsonapi]
 ActionDispatch::Http::Parameters::DEFAULT_PARSERS[media_type] = lambda do |body|
  data = JSON.parse(body)
  data = { :_json => data } unless data.is_a?(Hash)
  data.with_indifferent_access
end
ActionController::Renderers.add :jsonapi do |json, options|
  json = serialize_jsonapi(json, options).to_json(options) unless json.is_a?(String)
  self.content_type ||= media_type
  headers.merge! HEADERS[:response]
  self.response_body = json
end
```

## Configuring how a given renderer serializes/formats the model

e.g. the JSON renderer is defined as 

```ruby
    ActionController::Renders.add :json do |json, options|
      json = json.to_json(options) unless json.kind_of?(String)

      if options[:callback].present?
        if content_type.nil? || content_type == Mime[:json]
          self.content_type = Mime[:js]
        end

        "/**/#{options[:callback]}(#{json})"
      else
        self.content_type ||= Mime[:json]
        json
      end
    end
```

But I want to customize how my `json` object is serialized, and don't want to change the behavior of `to_json` in my model.

For example, in development I want to pretty-print my JSON responses. 

I should only need to change 
  - `json.to_json(options)` to 
  - `JSON.pretty_generate(json.as_json(options), options)`, 
but the current interface requires me to:

```diff
+  ActionController::Renderers.remove :json
  ActionController::Renders.add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+     unless json.is_a?(String)
+       json = json.as_json(options) if json.respond_to?(:as_json)
+       json = JSON.pretty_generate(json, options)
+     end

      if options[:callback].present?
        if content_type.nil? || content_type == Mime[:json]
          self.content_type = Mime[:js]
        end

        "/**/#{options[:callback]}(#{json})"
      else
        self.content_type ||= Mime[:json]
        json
      end
    end
```

This seems overkill when we're not changing the `response.content_type = Mime[:json]` (aka format) nor any other 'renderer' behavior.  We really just want to hook into the RHS of `response.response_body =`.

The [original implementation of ActiveModelSerializers back in 2011](https://github.com/rails/rails/commit/8ff7693a8dc61f43fc4eaf72ed24d3b8699191fe) was essentially at the same touch point: `json = json.active_model_serializer.new(json, serialization_scope) if json.respond_to?(:active_model_serializer)`.

### Updated Proposal

Over in https://github.com/rails/rails/pull/21496#discussion_r53316942 I just noticed that @dhh posted an interface that I really like.

```ruby
class ApplicationController < ActionController::Base
  # Takes a hash of form { 'renderer format' => 'callable that returns the response body' }
  serializing json: ->(object, options) { ActiveModel::SerializableResource.new(model, options) }
```

The Renderer would call something like
-  `serializing[:json].call(object, options)` 
- that defaults to `serializers[:json] = ->(object, options) { object.kind_of?(String) ? object : json.to_json(options) }`, 
- but can be customized and inherited by controllers:



I'm guessing that

> Here's a simple proposal for cutting down on the specialness of the configuration as much as possible

means that there would be a `class_attribute :serializing` and that the diff should be essentially:

```diff
+  class_attribute :serializing, instance_predicate: false
+  self.serializing = Hash.new do |h, format| 
+    fail ActionController::MissingSerializer, "There is no '#{format}' serializer.\n" <<
+                "Known serializers are #{serializing.keys}"
+  end
+  serializing[:json] = ->(object, options) { object.kind_of?(String) ? object : json.to_json(options) }
  ActionController::Renders.add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+     json = serializing[:json].call(json, options)

      if options[:callback].present?
        if content_type.nil? || content_type == Mime[:json]
          self.content_type = Mime[:js]
        end

        "/**/#{options[:callback]}(#{json})"
      else
        self.content_type ||= Mime[:json]
        json
      end
    end
```

bf4 commented on actionpack/CHANGELOG.md:
```diff
@@ -18,6 +18,24 @@
 
     *Greg Ose & Ben Toews*
 
+*   Add `ActionController::Renderers::SERIALIZERS` to separate rendering from serialization.
```
I'm wondering if `adapters` or `formatters` might be a better name here than `serializers`

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -145,19 +177,137 @@ def render_to_body(options)
-          method_name = Renderers._render_with_renderer_method_name(name)
-          return send(method_name, options.delete(name), options)
+      _renderers.each do |renderer_name|
+        next unless options.key?(renderer_name)
```
I wonder if there's a particular reason it doesn't check `formats` here, so that `render @model` with `formats: [:json]` would behave like `render json: @model` does now. Maybe that's only relevant in ActionController::MimeResponds::ClassMethods.`respond_to` blocks.

bf4 replied:
possibly related: https://github.com/rails/rails/pull/19378

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -145,19 +177,137 @@ def render_to_body(options)
+    add :json do |json, options|
       if options[:callback].present?
         if content_type.nil? || content_type == Mime[:json]
           self.content_type = Mime[:js]
```
with exception of the JSONp behavior, the renderer basically has three responsibilities

1. name the renderer for a format, e.g. `render json: @model` calls the `json` renderer
1. render the `response_body` (though it is set elsewhere)
1. set the `content_type` (mime type)

What if the renderer were more like

```ruby
Renderer = Struct.new(:content_type, :formatter)
```

So that `add_renderer :json` would basically `RENDERERS[:json][:json_api] = Renderer.new(Mime[:json_api], Proc.new do |object, options| { JsonApiResource.new(object, options).to_json end )`

and `render json: @model` would get to `_render_to_body_with_renderer` would do something like `renderer = RENDERERS[:json].find {|r| r.in?(formats) }; self.content_type = renderer.content_type, renderer.formatter.call(renderer_target_value, options)`

rafaelfranca commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -25,9 +42,16 @@ module Renderers
+    # A Hash mapping serializer names to callables.
+    # TODO: update text
+    # Default serializers are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
+    SERIALIZERS = Hash.new.with_indifferent_access
```
Why is it a Hash with indifferent access? We can use a normal hash here since we can define the API for how we add new serializers.

bf4 replied:
At this time I really don't remember.  I agree with what you wrote.

bf4 replied:
Ah, it was because in `_render_to_body_with_renderer`,  `serializer_name` is a String, even when passed in as a Symbol.

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -89,6 +120,57 @@ def self.remove(key)
+    # Example usage:
+    #
+    # Prior to the introduction of SERIALIZERS, customizing JSON serialization would
+    # have relied upon defining +_render_with_renderer_json+ (or +_render_option_json+,  pre-4.2 )
```
ref: https://github.com/rails/rails/commit/4dfe140ef3fbed34e6ae8af3b32402630a4690a1

bf4 committed:
[WIP] fleshing out DHH serializing proposal

https://github.com/rails/rails/pull/21496#discussion_r53316942

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -48,12 +76,18 @@ module All
     #   ActionController::Renderers.add :csv do |obj, options|
     #     filename = options[:filename] || 'data'
-    #     str = obj.respond_to?(:to_csv) ? obj.to_csv : obj.to_s
+    #     str = _serialize_with_serializer_csv(obj, options)
```
:scissors:

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -89,6 +123,68 @@ def self.remove(key)
+    # more background information.
+    def self.add_serializer(key, &block)
+      SERIALIZERS[key.to_sym] = block
+    end
```
- [ ] TODO: keep this interface? Since `serializing` is only availabe in the controller class, we need something at the module level for config, I think.  baby steps

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -145,19 +248,24 @@ def render_to_body(options)
+        renderer_target_method_name = Renderers._render_with_renderer_method_name(renderer_name)
+        renderer_target_value = options.delete(renderer_name)
+        serializer_name = options.key?(:serializer_name) ? options.delete(:serializer_name).to_sym : renderer_name
+        serialized_value = _serializers[serializer_name].call(renderer_target_value, options)
```
` _render_to_body_with_renderer` 

1. If there is a renderer for the format
1. Get the renderer method name and object to render
1. :new: get the serializer_name from options else same name as renderer/format
1. :new: serialize object using serializer with that name
1. send serialized object to renderer method (:new: which no longer modifies the object)

bf4 commented on actionpack/test/controller/renderers_test.rb:
```diff
@@ -86,5 +80,138 @@ def test_adding_csv_rendering_via_renderers_add
+    ActionController::Renderers.remove_serializer :csv
+  end
+
+  # FIXME: bundle exec ruby -Ilib:test test/controller/renderers_test.rb--seed 37264
```
:boom:

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -18,16 +18,33 @@ def initialize(format)
+
+    class_methods do
+      def serializing(**formatters)
+        self._formatters = _formatters.merge(formatters)
```
@dhh @rafaelfranca I implemented `serializing` to use a `_formatters` class_attribute which maps a format to its formatter.  (avoiding the word 'serializer'). (per https://github.com/rails/rails/pull/21496#discussion_r53316942 )

I've chosen to implement it in the least invasive way, which is to add a 'formatter' to the json/js/xml rendrerers, but not require it for the renderer to work, since the response body is set from the return value of the renderer in any case.

Other thoughts I've had are to make the 'RENDERERS' behave like the 'FORMATTERS', since formatters are 1->1. (but renderers are 1->0,1), but that makes some changes such as `use_renderer` more difficult.

I also considered adding the 'formatter' as a second argument to the renderer, but there wasn't a big advantage to that.

I'm interested in your thoughts this implementation, interface, and testing.

At this time, it is possible to add a formatter with no corresponding renderer. It just won't be used.  There's also not a clear public interface for removing a 'formatter'.  I'm not sure there should be one.

rafaelfranca replied:
> There's also not a clear public interface for removing a 'formatter'. I'm not sure there should be one.

Is not what `remove_formatter` does?

rafaelfranca replied:
I see `remove_formatter` only changes the default value. I think we are fine without one.

bf4 commented on actionpack/test/controller/renderers_test.rb:
```diff
@@ -87,4 +101,33 @@ def test_adding_csv_rendering_via_renderers_add
+    assert_equal "c,s,v", @response.body
+  ensure
+    ActionController::Renderers.remove :csv
+    TestController._formatters.delete(:csv)
```
Consider a more aggressive way to ensure all formatters of a format are removed from all serializers like how all renderers can be removed

rafaelfranca replied:
Can it use the `remove_formatter` API?

rafaelfranca replied:
Answering myself, no. `remove_formatter` changes the default value.

bf4 commented:
@rafaelfranca good to merge since @dhh said it looks fine? or add docs/changelog first? ( https://github.com/rails/rails/pull/21496#discussion_r61989150 ) or improve how formatters are removed? ( https://github.com/rails/rails/pull/21496#discussion_r62078729 )

maclover7 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -89,6 +107,15 @@ def self.remove(key)
+    def self.remove_formatter(format)
+      FORMATTERS.delete(format.to_sym)
+    end
+
```
✂️ extra line here

bf4 replied:
removed https://github.com/rails/rails/pull/21496/commits/98c26849d69f72648de39e587353885d593cd628

maclover7 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -18,16 +18,33 @@ def initialize(format)
     # A Set containing renderer names that correspond to available renderer procs.
     # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
     RENDERERS = Set.new
+    FORMATTERS = Hash.new do |h, format|
```
Is there a way to double check that we don't have duplicate format processors -- ex: should we prevent there from being 2 JSON formatters? I know we use a `Set` for `_renderers` 😬

bf4 replied:
In any given controller, there can be one and only one formatter for the `:json` format.

In general, the keys of the FORMATTERS Hash should always equal the elements in the RENDERERS Set (which is a Hash under the covers, anyhow).  

Or do you see something here I don't?

I do think there should be some refactoring to follow this up, but given all the back and forth discussion, chose the least-invasive implementation.


tl;dr summary of changes

```diff
# module Renderers

   included do
       class_attribute :_renderers
       self._renderers = Set.new.freeze
+      class_attribute :_formatters
+      self._formatters = FORMATTERS.dup.freeze
+    end

class_methods do
+      def serializing(**formatters)
+        self._formatters = _formatters.merge(formatters)
+      end
```

```diff
# module Renderers::All

      included do
         self._renderers = RENDERERS
+        self._formatters = FORMATTERS
       end
```

```diff
# module ActionController
+    def self.add_formatter(format, &block)
+      FORMATTERS[format.to_sym] = block
+    end
+

+    add_formatter :json do |json, options|
+      json.kind_of?(String) ? json : json.to_json(options)
+    end
     add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+      json = _formatters[:json].call(json, options)
```

rafaelfranca commented:
Implementation is great for me. Let's work in documentation and CHANGELOG.

bf4 commented:
[Test failure is in ActionCable, is unrelated](https://travis-ci.org/rails/rails/jobs/132136682)

bf4 commented:
@rafaelfranca @maclover7 Updated with small refactor and first go at docs and changelog. I'm certain there's, at the very least, style issues in the documentation and changelog.   I'm also adding new commits per feedback, but will rebase before completing.

rafaelfranca commented on guides/source/action_controller_overview.md:
```diff
@@ -666,6 +668,128 @@ end
+
+Note that this would only affect the +#index+ action.
+
+Two, by knowing that when the +JSON+ renderer is added (<tt>ActionController::Renderers.add :json, &block</tt>),
```
This is talking about an implementation detail and something that we consider private API, so I think it is better to not add this section to the documentation.

bf4 replied:
@rafaelfranca interesting.  When I talked to @wycats at RailsConf, I believe he said that he had intended `_render_with_renderer_json` as *the* public API, and that he had marked it as such.  This is the method (and its predecessor `_render_option_json`) used by ActiveModelSerializers.

rafaelfranca commented on guides/source/action_controller_overview.md:
```diff
@@ -666,6 +668,128 @@ end
+
+See <tt>ActionController::Renderers.use_renderers</tt> for more information.
+
+Should any of this go in guides/source/layouts_and_rendering.md ?
```
In fact I'd say this whole section should be there. This guide should only be a overview and we are going to deep on it.

bf4 commented on guides/source/action_controller_overview.md:
```diff
@@ -666,6 +668,128 @@ end
+Note that this would only affect the +#index+ action.
+
+Two, by knowing that when the +JSON+ renderer is added (<tt>ActionController::Renderers.add :json, &block</tt>),
+that the +_render_with_renderer_json+ method is generated, and knowing that the default Renderer does not modify strings,
```
@rafaelfranca interesting.  When I talked to @wycats at RailsConf, I believe he said that he had intended `_render_with_renderer_json` as *the* public API, and that he had marked it as such.  This is the method (and its predecessor `_render_option_json`) used by ActiveModelSerializers.

bf4 commented:
Failure is [ActionCable](https://travis-ci.org/rails/rails/jobs/132933944) and not related to this PR. PR is ready to be merged pending guides update approval

bf4 commented:
@rafaelfranca any chance this will be in 5.0.0?

bf4 commented:
Failure not related https://travis-ci.org/rails/rails/jobs/157761951

```
  log writing failed. PG::TRDeadlockDetected: ERROR:  deadlock detected
DETAIL:  Process 7399 waits for ShareLock on transaction 10049; blocked by process 7397.
Process 7397 waits for ShareLock on transaction 10048; blocked by process 7399.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (0,2) in relation "samples"
: UPDATE "samples" SET "value" = $1 WHERE "samples"."id" = $2
.

No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.

The build has been terminated
```

bf4 committed:
Merge branch 'master' into serializer_for_renderer

Conflicts:
	actionpack/CHANGELOG.md
	actionpack/lib/action_controller/metal/renderers.rb
	actionpack/test/controller/render_json_test.rb
	actionpack/test/controller/renderer_test.rb
	actionpack/test/controller/renderers_test.rb

bf4 committed:
Corrections per rubocop

aCandidMind commented:
This really seems to be the right direction and is looking like a good addition to make Rails a solid API choice and driving correct usage of the MIME type for json:api. It's only missing documentation & changelog approval so it seems to be upcoming.

bf4 commented on actionpack/lib/action_controller/metal/renderers.rb:
```diff
@@ -88,6 +120,68 @@ def self.remove(key)
+    #
+    #       json = json.as_json(options) if json.respond_to?(:as_json)
+    #       JSON.pretty_generate(json, options)
+    #     end
```
👀 Example

bf4 committed:
Merge branch 'master' into serializer_for_renderer

Conflicts:
	actionpack/CHANGELOG.md

bf4 committed:
Don't modify frozen string

bf4 committed:
Refactor naming in _render_to_body_with_renderer

bf4 committed:
Global SERIALIZERS/RENDERERS config needs to cascade to loaded descendants

bf4 committed:
Should probably dup RENDERERS Set, rather than make a new Set

bf4 committed:
Highlight todo

bf4 committed:
~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1726:in `current_server': DRb::DRbServerNotFound (DRb::DRbConnError)

```
rails/actionpack

📝 $ bundle exec ruby -Ilib:test test/controller/renderers_test.rb --seed 41487
Run options: --seed 41487

Running:

.....
~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1726:in `current_server': DRb::DRbServerNotFound (DRb::DRbConnError)
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1795:in `to_id'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1103:in `initialize'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:651:in `new'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:651:in `make_proxy'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:568:in `rescue in dump'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:565:in `dump'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:612:in `block in send_request'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:611:in `each'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:611:in `send_request'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:926:in `send_request'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1253:in `send_message'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1142:in `block (2 levels) in method_missing'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1229:in `open'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1141:in `block in method_missing'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1160:in `with_friend'
        from ~/.rvm/rubies/ruby-2.4.1/lib/ruby/2.4.0/drb/drb.rb:1140:in `method_missing'
        from rails/actionpack/test/abstract_unit.rb:404:in `block (2 levels) in shutdown'
        from rails/actionpack/test/abstract_unit.rb:393:in `fork'
        from rails/actionpack/test/abstract_unit.rb:393:in `block in shutdown'
        from rails/actionpack/test/abstract_unit.rb:392:in `times'
        from rails/actionpack/test/abstract_unit.rb:392:in `each'
        from rails/actionpack/test/abstract_unit.rb:392:in `map'
        from rails/actionpack/test/abstract_unit.rb:392:in `shutdown'
        from rails/bundle/ruby/2.4.0/gems/minitest-5.10.3/lib/minitest.rb:140:in `run'
        from rails/bundle/ruby/2.4.0/gems/minitest-5.10.3/lib/minitest.rb:63:in `block in autorun'
.

Finished in 0.056913s, 105.4235 runs/s, 245.9882 assertions/s.
6 runs, 14 assertions, 0 failures, 0 errors, 0 skips
```

bf4 committed:
Allow runtime changing of controller serializers

bf4 committed:
Disallow instance writing _serializers

bf4 committed:
Add actionpack/CHANGELOG to document use_renderers

bf4 committed:
Make test description/render format more explicit

bf4 committed:
Test explicit undefined renderer raises MissingTemplate

bf4 committed:
Remove unnecessary self.response_body =

bf4 committed:
Implement minimal ActionController.add/remove_serializer

bf4 committed:
Raise MissingSerializer when only Renderer defined

bf4 committed:
Assert exception messages

bf4 committed:
Add Controller.serializing mime: ->(obj, options) { }

Isolate JSON serializing tests from class_attribute effects

bf4 committed:
Merge branch 'master' into serializer_for_renderer

bf4 committed:
Make even with master to start impl over

bf4 committed:
Refactor _render_to_body_with_renderer

bf4 committed:
Improve naming in _render_to_body_with_renderer

bf4 committed:
Add changelog

bf4 commented:
@rafaelfranca below are the relevant commits to review.  I made a  kind of clean slate within the PR. You can see some of the commits are refactors that aren't strictly necessary.

| Message | ref |
|----------|-------|
| Merge branch 'master' into serializer_for_renderer | 4e7fdf0
| Make even with master to start impl over | c848569
| Add actionpack/CHANGELOG to document use_renderers | 1f06aa7
| Make test description/render format more explicit | 596179a
| Test explicit undefined renderer raises MissingTemplate | e6091fc
| Remove unnecessary self.response_body = | d305814
| Implement minimal ActionController.add/remove_serializer | a158880
| Raise MissingSerializer when only Renderer defined | c124ffb
| Assert exception messages | 1e5b552
| Add Controller.serializing mime: ->(obj, options) { } … | 70b2f85
| Refactor _render_to_body_with_renderer | cd00431
| Improve naming in _render_to_body_with_renderer | 44a68f0
| Add changelog | 4093649

Things that I think need attention

- [ ] Any refactors or changes to remove?

- [ ] That all the comments and documentation make sense, have no typos, are useful

- [ ] Consider adding a rescue_response to actionpack/lib/action_dispatch/middleware/exception_wrapper.rb for MissngSerializer and MissingRenderer

- [ ] Would be good to allow specifying the serializer as an argument to `render`, e.g.


```diff
-        serialized_value = _serializers[renderer_format].call(value_to_render, options)
+        serializer_name = (options.delete(:serializer_name) || renderer_format).to_sym
+        serialized_value = _serializers[serializer_name].call(value_to_render, options)
```

and

```diff
--- a/actionpack/CHANGELOG.md
+++ b/actionpack/CHANGELOG.md
@@ -16,11 +16,16 @@
     Renderers now only handle non-serialization concerns, such as setting the
     mime-type, how the data is returned, and handling callbacks.
 
-    The `serializer_name` is the name of the format being rendered, e.g. `json`.
+    The `serializer_name` is either from the params `serializer_name`, when given, or
+    the name of the format being rendered, e.g. `json`.
 
     The serializer is a callable object that accepts the same arguments as the renderer,
-    i.e. the object and the rendering options.  For example,
-    `render json: model` will serialize `model` as `_serializers[:json].call(model, options)`.
+    i.e. the object and the rendering options.
+        render json: model`
+        # => `_serializers[:json].call(model, options)
+
+        render json: model, serializer_name: :json_api
+        # =>  `_serializers[:json_api].call(model,options)
 
     The controller with raise an `ActionController::MissingSerializer` if no serializer is found
     for the format being rendered.
```

- [ ] Confirm there are no thread-safety issues of runtime errors. Tests would probably go in actionpack/test/controller/metal/renderers_test.rb

- [ ] Review interface and recommended usage, `ActionController.add_renderer` vs. `ActionController::Renderers.add`, etc.

- [ ] Compare `_renderers`/`use_renderer` to `_serializers`/`serializing` that they make sense to exist as implemented.

- [ ] Consider having the renderer `redefine` the method, rather than define it.

- [ ] Consider raising MissingRenderer in Rails itself, not just in the responders gem. 


```diff
--- a/actionpack/lib/action_controller/metal/renderers.rb
+++ b/actionpack/lib/action_controller/metal/renderers.rb
@@ -240,7 +240,9 @@ def serializing(options)
     # If no renderer is found, +super+ returns control to
     # <tt>ActionView::Rendering.render_to_body</tt>, if present.
     def render_to_body(options)
-      _render_to_body_with_renderer(options) || super
+      _render_to_body_with_renderer(options) \
+        || super \
+        || fail(ActionController::MissingRenderer.new(request.format))
     end
```

bf4 commented:
oh well
